### PR TITLE
Checking grammar

### PR DESCRIPTION
**Pull request by HoangNguyen679**

### CODE REVIEW

File: posts/rails_memory.md [hunk 1]
```diff
@@ -3,6 +3,8 @@ title: "Rails memory leak problems"
 date: "2021-10-10"
 ---
 
+Checking....
+
 Ruby memory leak problems are maybe familiar with Rubyist and you can find so many articles about them palso the solution with detailed explanation. As a Rails developer, I also stuck on that, and I want to share my experience. Let's start!
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - familiar with -> familiar to
  - somebody's familiar **with** something
  - something's familiar **to** somebody
- palso??
- detailed explanation -> detailed explanations or **a** detailed explanation
- I also stuck -> I'm stuck. Better way to say: I myself have also encountered this problem before

**Review by vannhi1118 — COMMENTED** [review]
Em chỉ comment được mỗi phần trước chỗ Problem showed up @@
Sau phần đó ko comment đc

File: posts/rails_memory.md [hunk 2]
```diff
@@ -3,14 +3,20 @@ title: "Rails memory leak problems"
...
 
+<!-- Checking.... -->
+
 Ruby memory leak problems are maybe familiar with Rubyist and you can find so many articles about them palso the solution with detailed explanation. As a Rails developer, I also stuck on that, and I want to share my experience. Let's start!
 
 # Problem showed up
 
+<!-- part 1 -->
+
 One morning, I woke up and received a message from the analysis team. They received 502 code when calling APIs. I started to investigate, of course, from server log. I went to Cloudwatch Log and found something were abnormal.
```
Comment (vannhi1118) [inline posts/rails_memory.md:11]: - found something were abnormal -> found something abnormal

File: posts/rails_memory.md [hunk 3]
```diff
@@ -3,14 +3,20 @@ title: "Rails memory leak problems"
...
 
 <img src="/images/high-memory-graph.png" />
 
+<!-- part 1.5 -->
+
 When analysis team called APIs, the memory percentage raise high, it is normal. But after that, the memory percentage did not decrease. And after 2-3 days, the memory went to 99%. That were unacceptable. And if I restart the server, it becames normal for the first day, after that, it's memory went up crazyly.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - When analysis team called APIs, the memory percentage raise high, it is normal -> It's normal if the memory percentage increases when APIs are called. 
- But after that, the memory percentage did not decrease. And after 2-3 days, the memory went to 99%. -> However, not only did it not decrease after that, 2-3 days later it went up to 99%. 
- That were unacceptable. -> That was unacceptable, or simply UNACCEPTABLE!, to emphasize how unacceptable it is :D
- if I restart the server, it becames normal for the first day, after that, it's memory went up crazyly. -> It became normal after I restarted the server, but just for the first day. After that, the memory went up crazily, again.

File: posts/rails_memory.md [hunk 4]
```diff
@@ -3,14 +3,20 @@ title: "Rails memory leak problems"
...
 
 At first, I thought maybe it were thread or process or gem memory leak problems. So I started to check.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - It were -> It was
- I thought maybe -> I assumed

**Review by vannhi1118 — COMMENTED** [review]
Em phải pack đồ lên đường đây nên comment đến đây thôi, lúc nào rảnh em comment tiếp =)))

File: posts/rails_memory.md [hunk 5]
```diff
@@ -19,24 +25,32 @@ At first, I thought maybe it were thread or process or gem memory leak problems.
 
 ## Derails
 
+<!-- part 2 -->
+
 I begined with [Derailed](https://github.com/zombocom/derailed_benchmarks) for general view of my Rails application. You can see memory and object created at required time with all gems the Rails app is using.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - I begined -> I began (begin is irregular verb, so its past and past participle form aren't simply adding "ed" at the end)
- Object -> objects

File: posts/rails_memory.md [hunk 6]
```diff
@@ -19,24 +25,32 @@ At first, I thought maybe it were thread or process or gem memory leak problems.
...
 And that report didn't seem wrong because I didn't found any issue with gem memory here.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - It's better not to use "And" to connect 2 sentences in writing. 
- And that report didn't seem wrong because I didn't found any issue with gem memory here. -> Nothing seemed wrong in the report because I didn't **find** any issue with gem memory here.
  - Find -> Past tense: Found, Did find, Didn't find

File: posts/rails_memory.md [hunk 7]
```diff
@@ -19,24 +25,32 @@ At first, I thought maybe it were thread or process or gem memory leak problems.
...
 
 You also can find the gem with memory problem [here](https://github.com/ASoftCo/leaky-gems)
 
 ## Puma
 
+<!-- part 3 -->
+
 So maybe Puma - the HTTP server of the app caused problem.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - So maybe **it's** **the Puma** - the HTTP server of the app **that** caused **the** problem.

File: posts/rails_memory.md [hunk 8]
```diff
@@ -19,24 +25,32 @@ At first, I thought maybe it were thread or process or gem memory leak problems.
...
 I received so many requests and use workers to handle them. Maybe Puma was sick and workers didn't work right. Actually, I found the [Puma memory problem](https://github.com/puma/puma/issues/342).
 
 People suggested that using [puma_worker_killer](https://github.com/zombocom/puma_worker_killer) but that was bad idea. My workers can be killed suddenly and the requests went to server can failed anytime without management. It was defenitely not the solution.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: You did not actually use the puma_worker_killer, you just made an assumption here, so it's better to use "would be" and "could" to emphasize the fact that you didn't implement this solution.
- People suggested ~that~ using [puma_worker_killer](https://github.com/zombocom/puma_worker_killer) but that ~was~ ** would be a** bad idea.
- My workers ~can~ could be killed suddenly and the requests went to **the** server ~can~ could fail~ed~ anytime without **additional**(?) management.

File: posts/rails_memory.md [hunk 9]
```diff
@@ -19,24 +25,32 @@ At first, I thought maybe it were thread or process or gem memory leak problems.
...
 
 ## Benchmark tools
 
+<!-- part 4 -->
+
 [puma_worker_killer](https://github.com/zombocom/puma_worker_killer) was not the solution. So if the problem at first, was not Puma ? I needed to confirm that. I have many requests so I would go to test the HTTP requests. You can use many tools to do that. Here I used wrk and siege.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - ~So~ **But what** if the problem, ~at first~ **originally**, was not **from** Puma ?
- You can use many tools to do that. Here I used wrk and siege. -> There are several tools that can perform this, but here I used wrk and siege.

File: posts/rails_memory.md [hunk 10]
```diff
@@ -65,13 +79,17 @@ CPU: 0.3% -> 7.5% -> 0.3%
 Memori: 9.7% -> 9.7% (no decrease)
 ```
 
+<!-- part 6 -->
+
 So when many requests came, the memory went high and not decrease. In the same time, CPU utilization went low after handling the requests.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - ~So~ **As you can see from the above result**, when many requests came, the memory went high and **did** not decrease. 
- In the same time, CPU utilization went low after handling the requests.
  - It should be **at** the same time. 
  - This phrase means something happens simultaneously. Here you're trying to explain sequential events, so you can use rephrase it to "After handling the requests, CPU utilization went low"

File: posts/rails_memory.md [hunk 11]
```diff
@@ -65,13 +79,17 @@ CPU: 0.3% -> 7.5% -> 0.3%
...
 
 But the APIs for the analysis team is inside AWS cloud, so I needed to check the internal requests.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: But the APIs ~for~ **used by** the analysis team is inside AWS cloud, so I needed to check the internal requests.
Write

File: posts/rails_memory.md [hunk 12]
```diff
@@ -65,13 +79,17 @@ CPU: 0.3% -> 7.5% -> 0.3%
...
 (I realized that if the public facing APIs requests went high, the server memory also went high, that's not good at all)
```
Comment (vannhi1118) [inline posts/rails_memory.md]: (I realized that if the public ~facing~ APIs requests went high, the server memory also went high, that's not good at all)

File: posts/rails_memory.md [hunk 13]
```diff
@@ -65,13 +79,17 @@ CPU: 0.3% -> 7.5% -> 0.3%
...
 
 ### Siege
 
+<!-- part 7 -->
+
 [Siege](https://github.com/JoeDog/siege) provides more useful options. So let's try with it.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - It's better to just combine these 2 sentences instead of separating them. 
- So let's try with it. -> So let's try it!

File: posts/rails_memory.md [hunk 14]
```diff
@@ -94,6 +112,8 @@ And called them
 siege -c10 -t30s -f urls.txt -H "Authorization: $authorization_key"
 ```
 
+<!-- part 8 -->
+
 Result was:
```
Comment (vannhi1118) [inline posts/rails_memory.md]: **The** result was **as follows**:

File: posts/rails_memory.md [hunk 15]
```diff
@@ -119,12 +139,16 @@ CPU: 0.03% -> 1.9% -> 0.05%
 Memory: 30% -> 39% -> 39%
 ```
 
+<!-- part 9 -->
+
 All my requests were handled but the memory did not decrease.
 It seemed like the Puma were not the root of the memory problem.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - It seemed like the Puma ~were~ **was** not the root of the memory problem.

**Review by vannhi1118 — COMMENTED** [review]
Em có task deadline hnay nên review nốt sau...

File: posts/rails_memory.md [hunk 16]
```diff
@@ -119,12 +139,16 @@ CPU: 0.03% -> 1.9% -> 0.05%
...
 I had to dig deeper and maybe it was ruby.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - I had to dig deeper and maybe it was ruby  -> I had to dig deeper, maybe the real culprit had been Ruby all along

File: posts/rails_memory.md [hunk 17]
```diff
@@ -139,6 +163,8 @@ RUN apt-get install -y libjemalloc-dev libjemalloc2
 ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2
 ```
 
+<!-- part 11 -->
+
 And, the memory was not high anymore. With the analysis team APIs called at around 15:00. It was totally OK.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: -> Finally the memory was not high anymore. It was totally OK when the analysis team called APIs (?) at around 15:00.

File: posts/rails_memory.md [hunk 18]
```diff
@@ -147,6 +173,8 @@ The server memory was often below 30% and I satisfied with that results.
 
 # More thinkings
```
Comment (vannhi1118) [inline posts/rails_memory.md]: - More thinkings -> Further discussion

File: posts/rails_memory.md [hunk 19]
```diff
@@ -147,6 +173,8 @@ The server memory was often below 30% and I satisfied with that results.
...
 
+<!-- part 12 -->
+
 You can find docker Ruby with jemalloc images [here](https://hub.docker.com/r/swipesense/ruby-jemalloc)
 It's quite convenient to start with the right choice of memory management library and it will mitigate your pain when discover the memory problems.
```
Comment (vannhi1118) [inline posts/rails_memory.md]: -> It's **recommended** to start with **a** right choice of memory management library **as** it will mitigate your pain **of encountering** memory problems.

**Review by vannhi1118 — COMMENTED** [review]
Ok xong hết rồi ạ =)))

File: posts/rails_memory.md (hunk 18 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/43385a881b5b9231888d65f8cdb96da739d3d352 fixed.

File: posts/rails_memory.md (hunk 19 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/43385a881b5b9231888d65f8cdb96da739d3d352 fixed.

File: posts/rails_memory.md (hunk 17 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/7ff21e904a7b15b9bca33fd3f4db95a0c855e513 fixed

File: posts/rails_memory.md (hunk 16 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/86d38bb661a525913e5ef5fbef28102c0c25fa93 fixed

File: posts/rails_memory.md (hunk 15 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/01006e5cdaac8f3517713ef877e0bad097b04a11 fixed

File: posts/rails_memory.md (hunk 14 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/01006e5cdaac8f3517713ef877e0bad097b04a11 fixed

File: posts/rails_memory.md (hunk 13 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/67e75a3f1569b4b89ae549039e123ca1ca8f8214 fixed

File: posts/rails_memory.md (hunk 12 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/914e37896fcd57ed34baf7f1fa62bfcb331cca6c fixed

File: posts/rails_memory.md (hunk 11 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/914e37896fcd57ed34baf7f1fa62bfcb331cca6c fixed

File: posts/rails_memory.md (hunk 10 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/914e37896fcd57ed34baf7f1fa62bfcb331cca6c fixed

File: posts/rails_memory.md (hunk 9 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/f6b68706a90f2671f0e0ac273aafaeec698a46ad fixed

File: posts/rails_memory.md (hunk 7 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/2dd104c77b6484c4dc36dc4339c84e3525b46d22 fixed

File: posts/rails_memory.md (hunk 8 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/2dd104c77b6484c4dc36dc4339c84e3525b46d22 fixed

File: posts/rails_memory.md (hunk 5 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/9da1aea95693cdc71204603a85c14b194debe761 fixed

File: posts/rails_memory.md (hunk 6 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/9da1aea95693cdc71204603a85c14b194debe761 fixed

File: posts/rails_memory.md (hunk 4 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/5304e5be3281d5c2a0e438f5efdab526f961704e fixed

File: posts/rails_memory.md (hunk 3 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/5304e5be3281d5c2a0e438f5efdab526f961704e fixed

File: posts/rails_memory.md (hunk 2 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md:11]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/5304e5be3281d5c2a0e438f5efdab526f961704e fixed

File: posts/rails_memory.md (hunk 1 above)
Comment (HoangNguyen679) [reply posts/rails_memory.md]: https://github.com/HoangNguyen679/aphelios/pull/1/commits/5304e5be3281d5c2a0e438f5efdab526f961704e fixed

**Comment by HoangNguyen679** [comment]
@vannhi1118 
Thanks nhaaa!
